### PR TITLE
Adding MANA counter and Promise joiner

### PR DESCRIPTION
**Pull request by evgenykuzyakov**

### CODE REVIEW

File: core/wasm/runtest/src/lib.rs [hunk 1]
```diff
@@ -43,27 +48,30 @@ impl External for MyExt {
...
-
-    fn promise_and(
-        &mut self,
-        _promise_id1: PromiseId,
-        _promise_id2: PromiseId,
-    ) -> ExtResult<PromiseId> {
-        Err(ExtError::NotImplemented)
+        match promise_id {
+            PromiseId::Receipt(_) => {
+                assert!(false);
```
Comment (bowenwang1996) [inline core/wasm/runtest/src/lib.rs]: Why do we need this assert? It already returns an error.

File: core/wasm/src/runtime.rs [hunk 2]
```diff
@@ -73,6 +79,27 @@ impl<'a> Runtime<'a> {
         }
     }
 
+    fn charge_mana(&mut self, mana: u32) -> bool {
+        let prev = self.mana_counter;
+        match prev.checked_add(mana) {
+            // mana charge overflow protection
+            None => false,
```
Comment (bowenwang1996) [inline core/wasm/src/runtime.rs]: The first two cases can be combined by `None | Some(val) if val > self.mana_limit`

File: core/wasm/runtest/src/lib.rs (hunk 1 above)
Comment (evgenykuzyakov) [reply core/wasm/runtest/src/lib.rs]: Easier to debug.

File: core/wasm/src/runtime.rs (hunk 2 above)
Comment (evgenykuzyakov) [reply core/wasm/src/runtime.rs]: This way is easier to read

**Pull request merged**